crates/criticaltrust/src/keys/public.rs

```diff
@@ -22,14 +23,51 @@ pub struct PublicKey {
 }
 
 impl PublicKey {
+    /// Verify whether the provided payload matches the provided signature and check if the
+    /// revoked content does not match the payload.
+    ///
+    /// Signature verification could fail if:
+    /// * The Expiration date has passed.
+    /// * The signature is present in the `RevocationInfo`.
+    /// * The `RevocationInfo` cannot be verified.
+    /// * [`verify_payload`](PublicKey::verify_without_revocations) fails.
+    pub fn verify(
+        &self,
+        role: KeyRole,
+        payload: &PayloadBytes<'_>,
+        signature: &SignatureBytes<'_>,
+        revocation_info: &RevocationInfo,
+    ) -> Result<(), Error> {
+        let current_utc = OffsetDateTime::now_utc();
+        let expiration_in_days = (revocation_info.expires_at - current_utc).whole_days();
+        if expiration_in_days <= 0 {
+            return Err(Error::SignaturesExpired);
+        }
+
+        let hashed_payload =
+            String::from_utf8_lossy(hash_sha256(payload.as_bytes()).as_slice()).to_string();
```

Comment:
I thought that `String::from_utf8_lossy` was ok here but I would like to know your opinions. This will only be called on the packages that are to be verified and can we assume those to be UTF8 strings?

Reply:
It might lead to a confusing error, but I also don't see this error popping up often.

I suggest adding to `errors.rs`:

```rust
    #[error(transparent)]
    FromUtf8(#[from] FromUtf8Error),
```

```suggestion
        let hashed_payload =
            String::from_utf8(hash_sha256(payload.as_bytes())).map_err(|e| Error::FromUtf8(e))?.to_string();
```

Reply:
This ended up getting fixed via reading the bytes from Sha256 as hex. Many thanks to @Hoverbear on help with this subtle bug.

```rust
let hashed_sha = hash_sha256(payload.as_bytes());
        let mut hashed_payload = String::new();
        for byte in hashed_sha {
            // The bytes returned by Sha256 (in hashed_sha) are to be read as hex and are not to be
            // confused with UTF-8 strings.
            hashed_payload.push_str(&format!("{byte:X}"));
        }
```